lib/fog/openstack/core.rb

```diff
@@ -169,13 +173,113 @@ def self.authenticate_v2(options, connection_options = {})
       }
     end
 
+    # Keystone Style Auth
+    def self.authenticate_v3(options, connection_options = {})
+      uri                   = options[:openstack_auth_uri]
+      tenant_name           = options[:openstack_tenant]
+      service_type          = options[:openstack_service_type]
+      service_name          = options[:openstack_service_name]
+      identity_service_type = options[:openstack_identity_service_type]
+      endpoint_type         = (options[:openstack_endpoint_type] || 'public').to_s
+      openstack_region      = options[:openstack_region]
+      domain_name           = options[:openstack_domain]
```

Comment:
My proposal user/domain split code for reference: https://github.com/fog/fog/pull/3518